cpp/test/sg/fil_test.cu

```diff
@@ -886,4 +889,139 @@ TEST_P(TreeliteThrowSparse8FilTest, Import) { check(); }
+  serial_multi_reduction(work, correct_result, p.set_size, p.n_sets);
+  T sum = multi_reduction<R>(work, p.set_size, p.n_sets);
+  if (threadIdx.x < p.set_size &&
+      1e-4 < fabsf(sum - correct_result[threadIdx.x])) {
+    atomicAdd(&error, 1);
+  }
+}
+
+template <typename T>
+__global__ void test_multi_reduction_k(T* data, MultiReductionTestParams* p,
+                                       int* error) {
+  int t = blockIdx.x;
+  switch (p[t].radix) {
+    case 2:
+      test_single_radix<2>(data[threadIdx.x], p[t], error[t]);
+      break;
+    case 3:
+      test_single_radix<3>(data[threadIdx.x], p[t], error[t]);
+      break;
+    case 4:
+      test_single_radix<4>(data[threadIdx.x], p[t], error[t]);
+      break;
+    case 5:
+      test_single_radix<5>(data[threadIdx.x], p[t], error[t]);
+      break;
+    case 6:
+      test_single_radix<6>(data[threadIdx.x], p[t], error[t]);
+      break;
+  }
+}
+
+template <typename T, void(generate_uniform)(T*, size_t)>
+class MultiReductionTest : public testing::TestWithParam<int> {
+ protected:
+  void SetUp() override {
+    block_dim_x = GetParam();
+    data_d.resize(block_dim_x);
+    T* data_p = data_d.data().get();
+    generate_uniform(data_p, block_dim_x);
+
+    for (int radix = 2; radix <= 6; ++radix) {
+      for (int set_size = 1; set_size < 15; ++set_size) {  // >2x the max radix
+        // 1..50 (if block_dim_x permits)
+        for (int n_sets = 1; n_sets <= std::min(block_dim_x, 50) / set_size;
+             ++n_sets)
+          p.push_back({.radix = radix, .set_size = set_size, .n_sets = n_sets});
+        // block_dim_x - 50 .. block_dim_x (if positive)
+        // up until 50 would be included in previous loop
+        for (int n_sets = std::max(block_dim_x - 50, 51) / set_size;
+             n_sets <= block_dim_x / set_size; ++n_sets)
+          p.push_back({.radix = radix, .set_size = set_size, .n_sets = n_sets});
+      }
+    }
+    p_d = p;
+    error_d.resize(p.size());
+    thrust::fill_n(error_d.begin(), p.size(), 0);
+  }
+
+  void check() {
+    T* data_p = data_d.data().get();
+    MultiReductionTestParams* p_p = p_d.data().get();
+    int* error_p = error_d.data().get();
+
+    test_multi_reduction_k<<<p.size(), block_dim_x>>>(data_p, p_p, error_p);
+    CUDA_CHECK(cudaPeekAtLastError());
+    error = error_d;
+    CUDA_CHECK(cudaDeviceSynchronize());
+    for (int i = 0; i < p.size(); ++i) {
+      ASSERT(error[i] == 0,
+             "test # %d: block_dim_x %d multi_reduction<%d>(on %d sets sized"
+             " %d) gave wrong result",
+             i, block_dim_x, p[i].radix, p[i].n_sets, p[i].set_size);
+    }
+  }
+
+  // parameters
+  raft::handle_t handle;
+  int block_dim_x;
+  thrust::host_vector<MultiReductionTestParams> p;
+  thrust::device_vector<MultiReductionTestParams> p_d;
+  thrust::host_vector<int> error;
+  thrust::device_vector<int> error_d;
+  thrust::device_vector<T> data_d;
+};
+
+std::vector<int> block_sizes = []() {
```

Comment:
Instead of going through all (even) block sizes, it's better to pick up a smaller number of values, e.g. typical values and corner cases, and initialize the vector with them.